File: python/infinicore/ops/mha_varlen.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def mha_varlen(
    q: Tensor,
    k: Tensor,
    v: Tensor,
    cum_seqlens_q: Tensor,
    cum_seqlens_k: Tensor,
    block_table: Tensor,
    max_seqlen_q: int,
    max_seqlen_k: int,
    alibi_slopes: Tensor | None = None,
    scale: float = 1.0,
    *,
    out: Tensor | None = None,
):
    if out is None:
        return Tensor(
            _infinicore.mha_varlen(
                q._underlying,
                k._underlying,
                v._underlying,
                cum_seqlens_q._underlying,
                cum_seqlens_k._underlying,
                block_table._underlying,
                max_seqlen_q,
                max_seqlen_k,
                alibi_slopes._underlying if alibi_slopes is not None else None,
                scale,
            )
        )

    _infinicore.mha_varlen_(
        out._underlying,
        q._underlying,
        k._underlying,
        v._underlying,
        cum_seqlens_q._underlying,
        cum_seqlens_k._underlying,
        block_table._underlying,
        max_seqlen_q,
        max_seqlen_k,
        alibi_slopes._underlying if alibi_slopes is not None else None,
        scale,
    )

    return out

File: csrc/infinicore/src/ops/multi_head_attention_varlen/mha_varlen.cc
#include "infinicore/ops/mha_varlen.hpp"
#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(MultiheadAttentionVarlen);

MultiheadAttentionVarlen::MultiheadAttentionVarlen(Tensor out,
                                                   const Tensor &q,
                                                   const Tensor &k,
                                                   const Tensor &v,
                                                   const Tensor &cum_seqlens_q,
                                                   const Tensor &cum_seqlens_kv,
                                                   std::optional<Tensor> block_table,
                                                   int max_seqlen_q,
                                                   int max_seqlen_k,
                                                   std::optional<Tensor> alibi_slopes,
                                                   float scale) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(out, q, k, v, cum_seqlens_q, cum_seqlens_kv);
    if (block_table.has_value()) {
        INFINICORE_ASSERT_TENSORS_SAME_DEVICE(out, block_table.value());
    }
    INFINICORE_GRAPH_OP_DISPATCH(out->device().type(),
                                 out, q, k, v, cum_seqlens_q, cum_seqlens_kv, block_table, max_seqlen_q, max_seqlen_k, alibi_slopes, scale);
}

void MultiheadAttentionVarlen::execute(Tensor out,
                                       const Tensor &q,
                                       const Tensor &k,
                                       const Tensor &v,
                                       const Tensor &cum_seqlens_q,
                                       const Tensor &cum_seqlens_kv,
                                       std::optional<Tensor> block_table,
                                       int max_seqlen_q,
                                       int max_seqlen_k,
                                       std::optional<Tensor> alibi_slopes,
                                       float scale) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(
        MultiheadAttentionVarlen,
        out, q, k, v, cum_seqlens_q, cum_seqlens_kv, block_table, max_seqlen_q, max_seqlen_k, alibi_slopes, scale);
}

Tensor mha_varlen(
    const Tensor &q,
    const Tensor &k,
    const Tensor &v,
    const Tensor &cum_seqlens_q,
    const Tensor &cum_seqlens_kv,
    std::optional<Tensor> block_table,
    int max_seqlen_q,
    int max_seqlen_k,
    std::optional<Tensor> alibi_slopes,
    float scale) {
    auto out_shape = q->shape();
    out_shape.back() = v->shape().back();
    auto out = Tensor::empty(out_shape, q->dtype(), q->device());
    mha_varlen_(out, q, k, v, cum_seqlens_q, cum_seqlens_kv, block_table, max_seqlen_q, max_seqlen_k, alibi_slopes, scale);
    return out;
}

void mha_varlen_(Tensor out,
                 const Tensor &q,
                 const Tensor &k,
                 const Tensor &v,
                 const Tensor &cum_seqlens_q,
                 const Tensor &cum_seqlens_kv,
                 std::optional<Tensor> block_table,
                 int max_seqlen_q,
                 int max_seqlen_k,
                 std::optional<Tensor> alibi_slopes,
                 float scale) {
    MultiheadAttentionVarlen::execute(out, q, k, v, cum_seqlens_q, cum_seqlens_kv, block_table, max_seqlen_q, max_seqlen_k, alibi_slopes, scale);
}

} // namespace infinicore::op
